### PR TITLE
Missing feature error 146

### PR DESCRIPTION
**Pull request by ejzn**
According to the spec feature is a required element during packaging. Currently the packager pukes since it tries to access an empty object for the Id. A check has now been added that standardizes the behaviour to return a proper error message. Fixes issue #146.

### CODE REVIEW

**Comment by jamesjhedley** [comment]
@erikj54 CI build is failing for this branch (jake test). Looks like you forgot to include the config-missing-feature.xml file. However, i suggest that we do not create/use this new file. Instead please mock the parsing (see below for similar example).

```
var data = testUtilities.cloneObj(testData.xml2jsConfig);

//Add a local feature element and a custom access list
data['feature'] = {'@': {id: 'blackberry.app', required: 'true', version: '1.0.0.0'}};//local feature
data['access'] = {'@': {uri: 'http://ci0000000094448.rim.net', subdomains: 'true'}};//custom access rule

mockParsing(data);
```

**Comment by ejzn** [comment]
Your right I did forget, my mistake. I saw that convention, here is my issue with it: I'm trying to test malformed/missing feature elements. So it doesn't really make sense to create one like the data. I guess one solution would be to create data['feature'] = undefined and data['feature'] = {}. Would this be the preferred way?

**Comment by jamesjhedley** [comment]
@erikj54 Or something like the below code should be sufficient, since by default "testData.xml2jsConfig" does not contain any features (i.e undefined).

```
var data = testUtilities.cloneObj(testData.xml2jsConfig);
mockParsing(data);
```

**Comment by tracyli** [comment]
tested in windows and mac with http://ci0000003863287:8080/hudson/job/BB10-Webworks-Packager-missing-feature-error/3/: this issue has been fixed.

**Comment by ejzn** [comment]
Right, what I should have said was an empty feature element. Feature by default is allowed to be undefined. Therefore the only valid test case is such that the feature is either not there, which is valid, but not needed to be tested by this test case, or is present, but empty, which is the goal of this test. I will make the changes and commit it.

File: lib/config-parser.js [hunk 1]
```diff
@@ -38,13 +38,15 @@ function createAccessListObj(featuresArray, uri, allowSubDomain) {
             allowSubDomain: allowSubDomain
         },
         attribs;
-
     if (featuresArray) {
         featuresArray.forEach(function (feature) {
+			//To assert that the feature has the proper Id which is required as per the spec
+            if (!feature["@"])
```
Comment (jamesjhedley) [inline lib/config-parser.js]: This should be a check() call. We should also move this logic and the id check below into the validateConfig function. That way all of our validation is in one place.

Comment (ejzn) [inline lib/config-parser.js]: That would make sense, and in fact that was my preferred way...except that if you check the initialization. The Parsing takes place before the validation. So the error showing up was being triggered during the initial parsing ie: processWidgetData() is what get's called at the beginning and was causing the failure. To accomplish this behaviour we would have to refactor a larger portion of this file.

**Comment by ishneur** [comment]
The error message is now: [ERROR]   Invalid element - failed to find the id attribute

It should actually be: [ERROR]   Invalid `<feature>` element - failed to find the id attribute

In my original issue I forgot to put quotes around `<feature>` so it wasn't visible in the expected error message, sorry. I fixed it now in blackberry/BB10-Webworks-Packager#146.

File: lib/config-parser.js (hunk 1 above)
Comment (jamesjhedley) [inline lib/config-parser.js]: Maybe i'm missing something, but cant we just do something like this? (see below)
If attribs wasn't set then we shouldn't try to access it or push it to the accessObj.

```
attribs = feature["@"];
if (attribs) {
    attribs.required = packagerUtils.toBoolean(attribs.required, true);
    accessObj.features.push(attribs);
}
```

Then we can move the validation logic u have created to the validateConfig function below.

File: lib/localize.js [hunk 2]
```diff
@@ -82,6 +82,9 @@ var Localize = require("localize"),
         "EXCEPTION_INVALID_ID": {
             "en": "Please enter a valid application id"
         },
+        "EXCEPTION_INVALID_FEATURE_ID": {
+            "en": "Invalid element - failed to find the id attribute"
```
Comment (jamesjhedley) [inline lib/localize.js]: This should probably read "Invalid feature - failed to find the id attribute"

File: lib/config-parser.js (hunk 1 above)
Comment (jamesjhedley) [inline lib/config-parser.js]: Lets try this;

```
attribs = feature["@"];
if (attribs) {
    attribs.required = packagerUtils.toBoolean(attribs.required, true);
}
accessObj.features.push(attribs);
```

And then further down in the validateConfig function you can insert the below code into the existing `widgetConfig.accessList.forEach()` loop.

```
//Check if there are empty features
if (access.features) {
    access.features.forEach(function (feature) {
        check(feature, localize.translate("EXCEPTION_INVALID_FEATURE_ID")).notNull();
        check(feature.id, localize.translate("EXCEPTION_INVALID_FEATURE_ID")).notNull();
    });
}
```

File: lib/localize.js [hunk 3]
```diff
@@ -82,6 +82,9 @@ var Localize = require("localize"),
         "EXCEPTION_INVALID_ID": {
             "en": "Please enter a valid application id"
         },
+        "EXCEPTION_INVALID_FEATURE_ID": {
+            "en": "Invalid feature - failed to find the id attribute"
```
Comment (jamesjhedley) [inline lib/localize.js]: Lets change this to the Snarf message provided by @ishneur

File: lib/localize.js [hunk 4]
```diff
@@ -82,6 +82,9 @@ var Localize = require("localize"),
         "EXCEPTION_INVALID_ID": {
             "en": "Please enter a valid application id"
         },
+        "EXCEPTION_INVALID_FEATURE_ID": {
+            "en": "Invalid <feature> - failed to find the id attribute"
```
Comment (jamesjhedley) [inline lib/localize.js]: Almost had it! :)

```
Invalid <feature> element - failed to find the id attribute
```

**Comment by jamesjhedley** [comment]
Code review complete. We should retest this, there were quite a few changes to the config parsing. @tracyli @ishneur

**Comment by tracyli** [comment]
tested in mac with http://ci0000003863287:8080/hudson/job/BB10-Webworks-Packager-missing-feature-error/9/: this issue has been fixed and the error message is "[ERROR] Invalid <feature> element - failed to find the id attribute" which is the same as in Snarf.

**Comment by nukulb** [comment]
merged manually

**Pull request closed without merging**